src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs

```diff
@@ -8,82 +8,69 @@
 using Microsoft.TemplateEngine.Abstractions;
 using Microsoft.TemplateEngine.Cli.CommandParsing;
 using Microsoft.TemplateEngine.Utils;
+using Microsoft.TemplateEngine.Cli.TemplateResolution;
+using System.Text;
 
 namespace Microsoft.TemplateEngine.Cli.HelpAndUsage
 {
     internal static class HelpForTemplateResolution
     {
-        public static CreationResultStatus CoordinateHelpAndUsageDisplay(TemplateListResolutionResult templateResolutionResult, IEngineEnvironmentSettings environmentSettings, INewCommandInput commandInput, IHostSpecificDataLoader hostDataLoader, ITelemetryLogger telemetryLogger, TemplateCreator templateCreator, string defaultLanguage, bool showUsageHelp = true)
+        public static CreationResultStatus CoordinateHelpAndUsageDisplay(ListOrHelpTemplateListResolutionResult templateResolutionResult, IEngineEnvironmentSettings environmentSettings, INewCommandInput commandInput, IHostSpecificDataLoader hostDataLoader, ITelemetryLogger telemetryLogger, TemplateCreator templateCreator, string defaultLanguage, bool showUsageHelp = true)
         {
             if (showUsageHelp)
             {
                 ShowUsageHelp(commandInput, telemetryLogger);
             }
 
-            // this is just checking if there is an unambiguous group.
-            // the called methods decide whether to get the default language filtered lists, based on what they're doing.
-            //
-            // The empty TemplateName check is for when only 1 template (or group) is installed.
-            // When that occurs, the group is considered partial matches. But the output should be the ambiguous case - list the templates, not help on the singular group.
-            if (!string.IsNullOrEmpty(commandInput.TemplateName)
-                    && templateResolutionResult.TryGetUnambiguousTemplateGroupToUse(out IReadOnlyList<ITemplateMatchInfo> unambiguousTemplateGroup)
-                    && TemplateListResolver.AreAllTemplatesSameLanguage(unambiguousTemplateGroup))
+            //in case list is specified we always need to list templates 
+            if (commandInput.IsListFlagSpecified)
             {
-                // This will often show detailed help on the template group, which only makes sense if they're all the same language.
-                return DisplayHelpForUnambiguousTemplateGroup(templateResolutionResult, environmentSettings, commandInput, hostDataLoader, templateCreator, defaultLanguage);
+                return DisplayListOrHelpForAmbiguousTemplateGroup(templateResolutionResult, environmentSettings, commandInput, hostDataLoader, telemetryLogger, defaultLanguage);
             }
-            else
+            else //help flag specified or no flag specified
             {
-                return DisplayHelpForAmbiguousTemplateGroup(templateResolutionResult, environmentSettings, commandInput, hostDataLoader, telemetryLogger, defaultLanguage);
+                if (!string.IsNullOrEmpty(commandInput.TemplateName)
```

Comment:
minor: would ```string::IsNullOrWhiteSpace``` better fit the test?
https://docs.microsoft.com/en-us/dotnet/api/system.string.isnullorwhitespace?view=netcore-3.1

Reply:
I don't think checking for whitespace is useful here, all whitespaces are already eliminated when parsing command line input.